File: src/services/eip681.ts
/**
 * EIP-681 payment-request URIs.
 * https://eips.ethereum.org/EIPS/eip-681
 *
 * We build the canonical forms below and parse them tolerantly (accepting the
 * legacy `pay-` prefix and scientific-notation amounts like `2.014e18`):
 *
 *   native, no amount   ethereum:<recipient>@<chainId>
 *   native + amount     ethereum:<recipient>@<chainId>?value=<wei>
 *   token, no amount    ethereum:<tokenAddr>@<chainId>/transfer?address=<recipient>
 *   token + amount      ethereum:<tokenAddr>@<chainId>/transfer?address=<recipient>&uint256=<baseUnits>
 *
 * This module is intentionally framework-free (no React / app-state imports) so
 * it can be unit-tested in isolation and reused from any layer.
 */

const ADDR_RE = /^0x[0-9a-fA-F]{40}$/;

export interface EIP681Request {
  /** Chain ID, or undefined when the URI omits `@chainId`. */
  chainId?: number;
  /** The address that should ultimately receive the funds. */
  recipient: string;
  /** ERC-20 contract address, or undefined for a native-coin payment. */
  tokenAddress?: string;
  /** Amount in the token's smallest unit (wei for native), or undefined if unspecified. */
  amountBaseUnits?: bigint;
  /** True for a native-coin payment, false for an ERC-20 `transfer`. */
  isNative: boolean;
}

export interface BuildEIP681Args {
  recipient: string;
  chainId: number;
  /** Omit (or null) for a native-coin request. */
  tokenAddress?: string | null;
  /** Decimals of the token (18 for native). Required when `amount` is set. */
  decimals?: number;
  /** Human-readable decimal amount, e.g. "1.5". Omit/empty for an open amount. */
  amount?: string;
}

export function isHexAddress(s: string): boolean {
  return ADDR_RE.test(s.trim());
}

/** Convert a human decimal string ("1.5") to integer base units given `decimals`. */
export function toBaseUnits(amount: string, decimals: number): bigint {
  const cleaned = (amount || '').trim();
  if (!cleaned) return 0n;
  const [intPart, fracPart = ''] = cleaned.split('.');
  const frac = (fracPart + '0'.repeat(decimals)).slice(0, decimals);
  return BigInt((intPart || '0') + frac);
}

/** Convert integer base units back to a trimmed human decimal string. */
export function fromBaseUnits(value: bigint, decimals: number): string {
  const s = value.toString();
  if (decimals <= 0) return s;
  const padded = s.padStart(decimals + 1, '0');
  const intPart = padded.slice(0, padded.length - decimals);
  const frac = padded.slice(padded.length - decimals).replace(/0+$/, '');
  return frac ? `${intPart}.${frac}` : intPart;
}

/**
 * Parse an EIP-681 numeric value (wei / uint256) into an integer bigint.
 * Accepts plain integers, decimals, and scientific notation (`2.014e18`).
 * Returns null for anything unparseable or negative.
 */
function parseAmount(raw: string): bigint | null {
  const m = raw.trim().match(/^([+-]?)(\d+)(?:\.(\d+))?(?:[eE]([+-]?\d+))?$/);
  if (!m) return null;
  const [, sign, intDigits, fracDigits = '', expStr = '0'] = m;
  if (sign === '-') return null;
  const digits = intDigits + fracDigits;
  const decimalExp = parseInt(expStr, 10) - fracDigits.length;
  let value: bigint;
  try {
    value = BigInt(digits);
  } catch {
    return null;
  }
  if (decimalExp >= 0) return value * 10n ** BigInt(decimalExp);
  // Negative exponent: truncate to an integer (base units are always integral).
  return value / 10n ** BigInt(-decimalExp);
}

/** Build a canonical EIP-681 URI for a receive request. */
export function buildEIP681(args: BuildEIP681Args): string {
  const { recipient, chainId, tokenAddress, decimals = 18, amount } = args;
  const hasAmount = !!amount && parseFloat(amount) > 0;

  if (tokenAddress) {
    let uri = `ethereum:${tokenAddress}@${chainId}/transfer?address=${recipient}`;
    if (hasAmount) uri += `&uint256=${toBaseUnits(amount!, decimals).toString()}`;
    return uri;
  }

  let uri = `ethereum:${recipient}@${chainId}`;
  if (hasAmount) uri += `?value=${toBaseUnits(amount!, decimals).toString()}`;
  return uri;
}

/**
 * Parse an EIP-681 URI. Returns null for anything that isn't a recognizable
 * `ethereum:` request (plain addresses, walletconnect/walletpair URIs, etc.),
 * so callers can fall through to their existing handling.
 */
export function parseEIP681(input: string): EIP681Request | null {
  let s = (input || '').trim();
  if (!/^ethereum:/i.test(s)) return null;
  s = s.slice('ethereum:'.length);
  if (s.startsWith('pay-')) s = s.slice('pay-'.length);

  // Split query string off.
  const qIndex = s.indexOf('?');
  const path = qIndex >= 0 ? s.slice(0, qIndex) : s;
  const query = qIndex >= 0 ? s.slice(qIndex + 1) : '';
  const params = parseQuery(query);

  // path = target[@chainId][/function]
  const slashIndex = path.indexOf('/');
  const targetWithChain = slashIndex >= 0 ? path.slice(0, slashIndex) : path;
  const functionName = slashIndex >= 0 ? path.slice(slashIndex + 1) : '';

  const atIndex = targetWithChain.indexOf('@');
  const target = (atIndex >= 0 ? targetWithChain.slice(0, atIndex) : targetWithChain).trim();
  const chainIdStr = atIndex >= 0 ? targetWithChain.slice(atIndex + 1) : '';

  if (!target) return null;
  const chainId = chainIdStr && /^\d+$/.test(chainIdStr) ? parseInt(chainIdStr, 10) : undefined;

  if (functionName === 'transfer') {
    // ERC-20: target is the token contract, recipient is the `address` param.
    const recipient = (params.address || '').trim();
    if (!isHexAddress(target) || !isHexAddress(recipient)) return null;
    const amountRaw = params.uint256 ?? params.value;
    return {
      chainId,
      recipient,
      tokenAddress: target,
      amountBaseUnits: amountRaw != null ? parseAmount(amountRaw) ?? undefined : undefined,
      isNative: false,
    };
  }

  // Native payment: target is the recipient. Accept only real addresses — ENS
  // resolution is out of scope and the send flow needs a 0x recipient anyway.
  if (!isHexAddress(target)) return null;
  return {
    chainId,
    recipient: target,
    amountBaseUnits: params.value != null ? parseAmount(params.value) ?? undefined : undefined,
    isNative: true,
  };
}

/** Public fallback host for payment links when there's no web origin (native app). */
export const PAY_LINK_FALLBACK = 'https://getvela.app/pay';

/**
 * The base for payment links. On the web we use the *current* origin's /pay
 * route, so a self-hosted wallet (e.g. mydomain.com) produces links that point
 * back to its own deployment. On native (no `window`) we fall back to the
 * public hosted page.
 */
export function payLinkBase(): string {
  try {
    if (typeof window !== 'undefined' && window.location?.origin) {
      return `${window.location.origin}/pay`;
    }
  } catch {
    /* not a browser */
  }
  return PAY_LINK_FALLBACK;
}

export interface PayLinkArgs {
  recipient: string;
  chainId: number;
  tokenAddress?: string | null;
  /** Human-readable decimal amount; omitted/zero → open amount. */
  amount?: string;
  /** Display hints so the landing page needs no RPC. */
  symbol: string;
  decimals: number;
  networkName: string;
  baseUrl?: string;
}

/**
 * Build a public, shareable payment-link URL — a web page (default getvela.app/pay)
 * that bridges to the Vela web wallet, to other EIP-681 wallets, or to manual entry.
 * The EIP-681 fields are carried as URL-encoded query params.
 */
export function buildPayLink(a: PayLinkArgs): string {
  const base = a.baseUrl ?? payLinkBase();
  const parts = [`to=${encodeURIComponent(a.recipient)}`, `chain=${a.chainId}`];
  if (a.tokenAddress) parts.push(`token=${encodeURIComponent(a.tokenAddress)}`);
  if (a.amount && parseFloat(a.amount) > 0) parts.push(`amount=${encodeURIComponent(a.amount)}`);
  parts.push(`sym=${encodeURIComponent(a.symbol)}`);
  parts.push(`dec=${a.decimals}`);
  if (a.networkName) parts.push(`net=${encodeURIComponent(a.networkName)}`);
  return `${base}?${parts.join('&')}`;
}

function parseQuery(query: string): Record<string, string> {
  const out: Record<string, string> = {};
  if (!query) return out;
  for (const pair of query.split('&')) {
    if (!pair) continue;
    const eq = pair.indexOf('=');
    const key = eq >= 0 ? pair.slice(0, eq) : pair;
    const val = eq >= 0 ? pair.slice(eq + 1) : '';
    try {
      out[decodeURIComponent(key)] = decodeURIComponent(val);
    } catch {
      out[key] = val;
    }
  }
  return out;
}

File: app-android/vela-wallet/app/src/main/java/app/getvela/wallet/feature/scan/Eip681.kt
package app.getvela.wallet.feature.scan

import java.math.BigInteger
import java.net.URLDecoder

/**
 * The EIP-681 tokenizer — the phone's port of the web's `services/eip681.ts`
 * (the desktop carries the same in Rust). It only tokenizes: whether a scan
 * locks the Send, whether a chainless request may lock at all, and how base
 * units become a figure are the core's (`send.rs::scan_resolved` /
 * `resolve_locked_request`). Two refusals bought with real bugs stay:
 *  - a call that is not a payment (`/approve`, …) is `null`, never a native
 *    send to the token contract;
 *  - in a `/transfer`, `value` is ether attached to the call, NOT the token
 *    amount — only `uint256` is.
 */
object Eip681 {
    class Request(
        val chainId: Long?,
        val recipient: String,
        val tokenAddress: String?,
        val amountBaseUnits: String?,
        val isNative: Boolean,
    )

    private val ADDRESS = Regex("^0x[0-9a-fA-F]{40}$")
    private val AMOUNT = Regex("^([+-]?)(\\d+)(?:\\.(\\d+))?(?:[eE]([+-]?\\d+))?$")

    fun isHexAddress(s: String): Boolean = ADDRESS.matches(s.trim())

    fun parse(input: String?): Request? {
        var s = input.orEmpty().trim()
        if (!s.startsWith("ethereum:", ignoreCase = true)) return null
        s = s.substring("ethereum:".length)
        if (s.startsWith("pay-")) s = s.substring("pay-".length)
        val q = s.indexOf('?')
        val path = if (q >= 0) s.substring(0, q) else s
        val query = if (q >= 0) s.substring(q + 1) else ""
        val params = parseQuery(query)
        val slash = path.indexOf('/')
        val targetWithChain = if (slash >= 0) path.substring(0, slash) else path
        val functionName = if (slash >= 0) path.substring(slash + 1) else ""
        val at = targetWithChain.indexOf('@')
        val target = (if (at >= 0) targetWithChain.substring(0, at) else targetWithChain).trim()
        val chainText = if (at >= 0) targetWithChain.substring(at + 1) else ""
        if (target.isEmpty()) return null
        val chainId = chainText.takeIf { it.isNotEmpty() && it.all(Char::isDigit) }?.toLongOrNull()
        if (functionName == "transfer") {
            val recipient = params["address"].orEmpty().trim()
            if (!isHexAddress(target) || !isHexAddress(recipient)) return null
            return Request(
                chainId = chainId,
                recipient = recipient,
                tokenAddress = target,
                amountBaseUnits = params["uint256"]?.let(::parseAmount),
                isNative = false,
            )
        }
        if (functionName.isNotEmpty()) return null
        if (!isHexAddress(target)) return null
        return Request(
            chainId = chainId,
            recipient = target,
            tokenAddress = null,
            amountBaseUnits = params["value"]?.let(::parseAmount),
            isNative = true,
        )
    }

    /** `1e18`, `2.5e6`, `1000`; a negative answers `null` (nothing), as the web. */
    internal fun parseAmount(raw: String): String? {
        val m = AMOUNT.matchEntire(raw.trim()) ?: return null
        val (sign, intDigits, fracDigits, expText) = m.destructured
        if (sign == "-") return null
        val exp = if (expText.isEmpty()) 0 else expText.toIntOrNull() ?: return null
        val digits = intDigits + fracDigits
        val decimalExp = exp - fracDigits.length
        if (decimalExp > 512) return null
        val value = BigInteger(digits)
        return if (decimalExp >= 0) {
            (value * BigInteger.TEN.pow(decimalExp)).toString()
        } else {
            value.divide(BigInteger.TEN.pow(-decimalExp)).toString()
        }
    }

    internal fun parseQuery(query: String): Map<String, String> {
        if (query.isEmpty()) return emptyMap()
        val out = LinkedHashMap<String, String>()
        query.split('&').forEach { pair ->
            if (pair.isEmpty()) return@forEach
            val eq = pair.indexOf('=')
            val key = if (eq >= 0) pair.substring(0, eq) else pair
            val value = if (eq >= 0) pair.substring(eq + 1) else ""
            out[decode(key)] = decode(value)
        }
        return out
    }

    private fun decode(raw: String): String =
        runCatching { URLDecoder.decode(raw.replace("+", "%2B"), "UTF-8") }.getOrDefault(raw)
}
